From d4ad2f60d264ed254ee169dff68e5903e5e24561 Mon Sep 17 00:00:00 2001
From: "Tan, Raymond" <raymond.tan@intel.com>
Date: Mon, 13 Apr 2015 11:31:02 +0800
Subject: [PATCH] platform: x86: Enable pdata for Intel GIP

This patch enable the pdata for GIP driver on Intel Quark X1000 SoC.
This was originally from a larger patch in Quark BSP R1.1., which
was authored by Josef Ahmad <josef.ahmad@linux.intel.com>.

Signed-off-by: Tan, Raymond <raymond.tan@intel.com>
Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>

diff --git a/drivers/platform/x86/intel-quark/intel_quark_platform_galileo.c b/drivers/platform/x86/intel-quark/intel_quark_platform_galileo.c
index b97e8a801ec3..bdf8eb1a3ece 100644
--- a/drivers/platform/x86/intel-quark/intel_quark_platform_galileo.c
+++ b/drivers/platform/x86/intel-quark/intel_quark_platform_galileo.c
@@ -35,6 +35,7 @@
 #include <linux/spi/pxa2xx_spi.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
+#include <linux/mfd/intel_qrk_gip_pdata.h>
 
 #define DRIVER_NAME "Galileo"
 
@@ -105,6 +106,15 @@ static struct cy8c9540a_pdata cy8c9540a_platform_data = {
 	.irq_base		= 64,
 };
 
+/* Cypress expander requires i2c master to operate @100kHz 'standard mode' */
+static struct intel_qrk_gip_pdata gip_pdata = {
+	.i2c_std_mode = 1,
+};
+static struct intel_qrk_gip_pdata *galileo_gip_get_pdata(void)
+{
+	return &gip_pdata;
+}
+
 /******************************************************************************
  *             Analog Devices AD7298 SPI Device Platform Data
  ******************************************************************************/
@@ -279,6 +289,9 @@ static int intel_quark_galileo_gen1_init(struct platform_device *pdev)
 	static int spi_done;
 	static int gpios_done;
 
+	/* Assign GIP driver handle for board-specific settings */
+	intel_qrk_gip_get_pdata = galileo_gip_get_pdata;
+
 	if (spi_done)
 		goto gpios;
 
diff --git a/drivers/platform/x86/intel-quark/intel_quark_platform_galileo_gen2.c b/drivers/platform/x86/intel-quark/intel_quark_platform_galileo_gen2.c
index a6a35bfa7e96..f201d1776de0 100644
--- a/drivers/platform/x86/intel-quark/intel_quark_platform_galileo_gen2.c
+++ b/drivers/platform/x86/intel-quark/intel_quark_platform_galileo_gen2.c
@@ -35,6 +35,7 @@
 #include <linux/spi/pxa2xx_spi.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
+#include <linux/mfd/intel_qrk_gip_pdata.h>
 
 /******************************************************************************
  *                        Intel Galileo Gen2
@@ -49,6 +50,15 @@ static int gpio_cs;
 module_param(gpio_cs, int, S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(gpio_cs, "Enable GPIO chip-select for SPI channel 1");
 
+/* Galileo Gen2 boards require i2c master to operate @400kHz 'fast mode' */
+static struct intel_qrk_gip_pdata gip_pdata_gen2 = {
+	.i2c_std_mode = 0,
+};
+static struct intel_qrk_gip_pdata *galileo_gen2_gip_get_pdata(void)
+{
+	return &gip_pdata_gen2;
+}
+
 /******************************************************************************
  *             Texas Instruments ADC1x8S102 SPI Device Platform Data
  ******************************************************************************/
@@ -266,6 +276,9 @@ static int intel_quark_galileo_gen2_init(struct platform_device *pdev)
 	struct i2c_adapter *i2c_adap = NULL;
 	struct i2c_client *client = NULL;
 
+	/* Assign GIP driver handle for board-specific settings */
+	intel_qrk_gip_get_pdata = galileo_gen2_gip_get_pdata;
+
 	/* Need to tell the PCA953X driver which GPIO IRQ to use for signalling
 	 * interrupts.  We can't get the IRQ until the GPIO driver is loaded.
 	 * Hence, we defer registration of the I2C devices until now
-- 
2.1.0

